src/ac_training_lab/A1-cam/device.py

```diff
@@ -0,0 +1,118 @@
+import paho.mqtt.client as mqtt
+import boto3
+import json
+from datetime import datetime
+import time
+
+from picamera2 import Picamera2
+from libcamera import controls, Transform
+
+from my_secrets import (
+    CAMERA_READ_ENDPOINT,
+    CAMERA_WRITE_ENDPOINT,
+    MQTT_PORT,
+    MQTT_HOST,
+    MQTT_PASSWORD,
+    MQTT_USERNAME,
+)
+
+import logging
+import sys
+
+logging.basicConfig(
+    level=logging.INFO,
+    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
+    handlers=[logging.StreamHandler(sys.stdout), logging.FileHandler("device.log")],
+)
+logger = logging.getLogger("device")
+
+
+def on_message(client, userdata, message):
+    data = json.loads(message.payload)
+    command = data["command"]
+
+    if command == "capture_image":
+        logging.info(f"Received command: {command}")
+
+        # only keeping one local copy at any given time called image.jpg
+        file_path = "image.jpeg"
+
+        logging.info("Begin autofocus")
+        try:
+            picam2.autofocus_cycle()
+            logging.info("Successfully finished autofocus")
+        except Exception as e:
+            logging.error(f"Error autofocusing: {e}")
+            return
+
+        logging.info("Start image capture")
+        try:
+            picam2.capture_file(file_path)
+            logging.info("Successfully capture image")
+        except Exception as e:
+            logging.error(f"Error capturing image: {e}")
+
+        # make sure to setup S3 bucket with ACL and public access so that the link works publically
+        # using the current timestamp as the unique object ID
+        object_name = datetime.utcnow().strftime("%Y-%m-%d-%H:%M:%S") + ".jpeg"
+        logging.info("Begin upload to S3")
+        try:
+            s3.upload_file(
```

Comment:
How is s3 being used prior to being defined? If it were a global variable, I'd understand, but I only see it being defined in the `if __main__` towards the end. Maybe I'm missing something basic.